Bug Report: VtBackup always aborts when encountering replication issue

### Overview of the Issue

When `vtbackup` encounters a replication issue, it will restart replication in an attempt to mitigate the issue. However, even if we get back to a healthy state, vtbackup will still abort the entire backup.


For example, if a failover happens while a backup is running,`vtbackup` will hit `Replica I/O for channel '': Error reconnecting to source '<user>@<ip>:3306'. This was attempt 1/300, with a delay of 10 seconds between attempts. Message: Can't connect to MySQL server on '<ip>:3306' (111), Error_code: MY-002003` and will execute the following statement:
```
STOP REPLICA;
CHANGE REPLICATION SOURCE TO
  SOURCE_HOST = '<ip 2>',  #<--- New host
  SOURCE_PORT = 0,
  SOURCE_USER = '<user>',
  SOURCE_PASSWORD = '****',
  ...
START REPLICA;
```

However, even if the failover is successful, and we return to a healthy state, 60s later, vtbacup will abort with the following error: 
```
I0513 22:01:17.399770       1 mysqld.go:436] Mysqld.Start(1778709440) stderr: 2026-05-13T22:01:17.399714Z 38 [System] [MY-014002] [Repl] Replica receiver thread for channel '': connected to source '<user>@<ip>:3306' with server_uuid=<uuid>, server_id=<id>. Starting GTID-based replication.
E0513 22:02:16.470488       1 last_error.go:85] replication catch up: the same error was encountered continuously since 2026-05-13 22:01:15.862994723 +0000 UTC, it is now assumed to be unrecoverable; any affected operations will need to be manually restarted once error 'replication has stopped before backup could be taken. trying to restart replication.' has been addressed
I0513 22:02:16.470545       1 mysqld.go:625] Mysqld.Shutdown
...
```

## Root Cause
Vtbackup uses a `LastError` at `go/cmd/vtbackup/cli/vtbackup.go:689`. Whenever it notices an error with the replication's health, it records the error. This happens at line `734`:
```GO
if !status.Healthy() {
  errStr := "Replication has stopped before backup could be taken. Trying to restart replication."
  log.Warn(errStr)
  lastErr.Record(errors.New(strings.ToLower(errStr)))
  ...
```
However, we never clear the error via `lastErr.Record(nil)` whenever we get back to a healthy state. Meaning that as soon as we hit an error once, regardless of if it was transient or not, the backup will abort in 60s.

### Reproduction Steps

1. Have a Vitess Cluster running
2. Start a VBJ
3. Kill the primary, triggering an ERS
4. Notice the following in the logs
```
I0513 22:01:15.101213       1 mysqld.go:436] Mysqld.Start(1778709440) stderr: 2026-05-13T22:01:15.100264Z 13 [ERROR] [MY-010557] [Repl] Error reading packet from server for channel '': Lost connection to MySQL server during query (server_errno=2013)
I0513 22:01:15.101281       1 mysqld.go:436] Mysqld.Start(1778709440) stderr: 2026-05-13T22:01:15.100320Z 13 [Warning] [MY-010897] [Repl] Storing MySQL user name or password information in the connection metadata repository is not secure and is therefore not recommended. Please consider using the USER and PASSWORD connection options for START REPLICA; see the 'START REPLICA Syntax' in the MySQL Manual for more information.
I0513 22:01:15.101292       1 mysqld.go:436] Mysqld.Start(1778709440) stderr: 2026-05-13T22:01:15.100668Z 13 [ERROR] [MY-010584] [Repl] Replica I/O for channel '': Error reconnecting to source '<user>@<ip 1>:3306'. This was attempt 1/300, with a delay of 10 seconds between attempts. Message: Can't connect to MySQL server on '<ip 1>:3306' (111), Error_code: MY-002003
W0513 22:01:15.862966       1 vtbackup.go:735] Replication has stopped before backup could be taken. Trying to restart replication.
I0513 22:01:15.873688       1 query.go:91] exec STOP REPLICA
I0513 22:01:16.064296       1 query.go:91] exec CHANGE REPLICATION SOURCE TO
  SOURCE_HOST = '<ip 2>',
  SOURCE_PORT = 0,
  SOURCE_USER = '<user>',
  SOURCE_PASSWORD = '****',
  SOURCE_CONNECT_RETRY = 10,
  GET_SOURCE_PUBLIC_KEY = 1,
  SOURCE_
I0513 22:01:16.066416       1 mysqld.go:436] Mysqld.Start(1778709440) stderr: 2026-05-13T22:01:16.066326Z 12 [System] [MY-010597] [Repl] 'CHANGE REPLICATION SOURCE TO FOR CHANNEL '' executed'. Previous state source_host='<ip 1>', source_port= 3306, source_log_file='<binlog>.000007', source_log_pos= 7081, source_bind=''. New state source_host='<ip 2>', source_port= 3306, source_log_file='', source_log_pos= 4, source_bind=''.
I0513 22:01:16.360647       1 query.go:91] exec START REPLICA
I0513 22:01:16.362450       1 mysqld.go:436] Mysqld.Start(1778709440) stderr: 2026-05-13T22:01:16.361194Z 36 [Warning] [MY-010897] [Repl] Storing MySQL user name or password information in the connection metadata repository is not secure and is therefore not recommended. Please consider using the USER and PASSWORD connection options for START REPLICA; see the 'START REPLICA Syntax' in the MySQL Manual for more information.
I0513 22:01:16.363470       1 mysqld.go:436] Mysqld.Start(1778709440) stderr: 2026-05-13T22:01:16.362574Z 36 [ERROR] [MY-010584] [Repl] Replica I/O for channel '': Error connecting to source '<user>@<ip 2>:3306'. This was attempt 1/300, with a delay of 10 seconds between attempts. Message: Can't connect to MySQL server on '<ip 2>:3306' (111), Error_code: MY-002003
W0513 22:01:17.364779       1 vtbackup.go:735] Replication has stopped before backup could be taken. Trying to restart replication.
I0513 22:01:17.372488       1 query.go:91] exec STOP REPLICA
I0513 22:01:17.376523       1 query.go:91] exec CHANGE REPLICATION SOURCE TO
  SOURCE_HOST = '<ip 2>',
  SOURCE_PORT = 0,
  SOURCE_USER = '<user>',
  SOURCE_PASSWORD = '****',
  SOURCE_CONNECT_RETRY = 10,
  GET_SOURCE_PUBLIC_KEY = 1,
  SOURCE_
I0513 22:01:17.376754       1 mysqld.go:436] Mysqld.Start(1778709440) stderr: 2026-05-13T22:01:17.376709Z 12 [System] [MY-010597] [Repl] 'CHANGE REPLICATION SOURCE TO FOR CHANNEL '' executed'. Previous state source_host='<ip 2>', source_port= 3306, source_log_file='', source_log_pos= 4, source_bind=''. New state source_host='<ip 2>', source_port= 3306, source_log_file='', source_log_pos= 4, source_bind=''.
I0513 22:01:17.394481       1 query.go:91] exec START REPLICA
I0513 22:01:17.395271       1 mysqld.go:436] Mysqld.Start(1778709440) stderr: 2026-05-13T22:01:17.395222Z 38 [Warning] [MY-010897] [Repl] Storing MySQL user name or password information in the connection metadata repository is not secure and is therefore not recommended. Please consider using the USER and PASSWORD connection options for START REPLICA; see the 'START REPLICA Syntax' in the MySQL Manual for more information.
I0513 22:01:17.399770       1 mysqld.go:436] Mysqld.Start(1778709440) stderr: 2026-05-13T22:01:17.399714Z 38 [System] [MY-014002] [Repl] Replica receiver thread for channel '': connected to source '<user>@<ip 2>:3306' with <uuid> server_id=247111263. Starting GTID-based replication.
E0513 22:02:16.470488       1 last_error.go:85] replication catch up: the same error was encountered continuously since 2026-05-13 22:01:15.862994723 +0000 UTC, it is now assumed to be unrecoverable; any affected operations will need to be manually restarted once error 'replication has stopped before backup could be taken. trying to restart replication.' has been addressed
I0513 22:02:16.470545       1 mysqld.go:625] Mysqld.Shutdown
```

### Binary Version

```sh
vtbackup version Version: 22.0.0-SNAPSHOT (Git revision 55b0e35e84c64747625b943c54460cc0562bc46f branch 'latest-22.0') built on Thu May 14 06:13:49 UTC 2026 by vitess@buildkitsandbox using go1.24.11 linux/amd64
```

### Operating System and Environment details

```sh
ID=wolfi
NAME="Wolfi"
PRETTY_NAME="Wolfi"
VERSION_ID="20230201"
HOME_URL="https://wolfi.dev"
BUG_REPORT_URL="https://github.com/wolfi-dev/os/issues"
Darwin 25.4.0
arm64
```

### Log Fragments

```sh
Provided above
```

--- a/go/cmd/vtbackup/cli/vtbackup.go
+++ b/go/cmd/vtbackup/cli/vtbackup.go
@@ -741,6 +741,8 @@ func catchUpReplicationForBackup(ctx context.Context, topoServer *topo.Server, m
 				log.Warn(fmt.Sprintf("Failed to restart replication: %v", err))
 			}
 		} else {
+			// Clear last error as MySQL's replication is healthy
+			lastErr.Record(nil)
 			phaseStatus.Set([]string{phaseNameCatchupReplication, phaseStatusCatchupReplicationStopped}, 0)
 		}
 	}

--- a/go/cmd/vtbackup/cli/vtbackup_test.go
+++ b/go/cmd/vtbackup/cli/vtbackup_test.go
@@ -1,0 +1,138 @@
+/*
+Copyright 2026 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cli
+
+import (
+	"context"
+	"testing"
+	"testing/synctest"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"vitess.io/vitess/go/mysql/replication"
+	"vitess.io/vitess/go/vt/mysqlctl"
+	"vitess.io/vitess/go/vt/topo"
+	"vitess.io/vitess/go/vt/topo/memorytopo"
+
+	topodatapb "vitess.io/vitess/go/vt/proto/topodata"
+)
+
+func TestCatchUpReplicationForBackupClearsLastErrWhenReplicationBecomesHealthy(t *testing.T) {
+	synctest.Test(t, func(t *testing.T) {
+		ctx := t.Context()
+
+		oldInitKeyspace := initKeyspace
+		oldInitShard := initShard
+		initKeyspace = "test_keyspace"
+		initShard = "0"
+		t.Cleanup(func() {
+			initKeyspace = oldInitKeyspace
+			initShard = oldInitShard
+		})
+
+		ts := memorytopo.NewServer(ctx, "zone1")
+		t.Cleanup(ts.Close)
+		require.NoError(t, ts.CreateKeyspace(ctx, initKeyspace, &topodatapb.Keyspace{}))
+		require.NoError(t, ts.CreateShard(ctx, initKeyspace, initShard))
+		primaryAlias := &topodatapb.TabletAlias{Cell: "zone1", Uid: 100}
+		require.NoError(t, ts.CreateTablet(ctx, &topodatapb.Tablet{
+			Alias:         primaryAlias,
+			Keyspace:      initKeyspace,
+			Shard:         initShard,
+			Hostname:      "primary.test",
+			MysqlHostname: "primary-mysql.test",
+			MysqlPort:     3306,
+			Type:          topodatapb.TabletType_PRIMARY,
+		}))
+		_, err := ts.UpdateShardFields(ctx, initKeyspace, initShard, func(si *topo.ShardInfo) error {
+			si.PrimaryAlias = primaryAlias
+			return nil
+		})
+		require.NoError(t, err)
+
+		restorePos := testCatchupPosition(1)
+		primaryPos := testCatchupPosition(3)
+		statuses := []replication.ReplicationStatus{
+			{
+				Position:    restorePos,
+				IOState:     replication.ReplicationStateConnecting,
+				LastIOError: "Replica I/O for channel '': Error reconnecting to source 'vt_test@192.0.2.10:3306'. This was attempt 1/300, with a delay of 10 seconds between attempts. Message: Can't connect to MySQL server on '192.0.2.10:3306' (111), Error_code: MY-002003",
+				SQLState:    replication.ReplicationStateRunning,
+			},
+		}
+		for range int(timeoutWaitingForReplicationStatus.Seconds()) + 1 {
+			statuses = append(statuses, replication.ReplicationStatus{
+				Position: restorePos,
+				IOState:  replication.ReplicationStateRunning,
+				SQLState: replication.ReplicationStateRunning,
+			})
+		}
+		statuses = append(
+			statuses,
+			replication.ReplicationStatus{
+				Position: primaryPos,
+				IOState:  replication.ReplicationStateRunning,
+				SQLState: replication.ReplicationStateRunning,
+			},
+			replication.ReplicationStatus{
+				Position: primaryPos,
+				IOState:  replication.ReplicationStateRunning,
+				SQLState: replication.ReplicationStateRunning,
+			},
+		)
+		mysqld := &catchupReplicationMysqlDaemon{
+			statuses: statuses,
+		}
+
+		status, err := catchUpReplicationForBackup(ctx, ts, mysqld, restorePos, primaryPos)
+
+		require.NoError(t, err)
+		assert.True(t, status.Position.Equal(primaryPos))
+		assert.Equal(t, 1, mysqld.setReplicationSourceCalls)
+		assert.Equal(t, 1, mysqld.stopReplicationCalls)
+	})
+}
+
+type catchupReplicationMysqlDaemon struct {
+	mysqlctl.MysqlDaemon
+
+	statuses                  []replication.ReplicationStatus
+	statusCalls               int
+	setReplicationSourceCalls int
+	stopReplicationCalls      int
+}
+
+func (m *catchupReplicationMysqlDaemon) ReplicationStatus(ctx context.Context) (replication.ReplicationStatus, error) {
+	status := m.statuses[m.statusCalls]
+	m.statusCalls++
+	return status, nil
+}
+
+func (m *catchupReplicationMysqlDaemon) SetReplicationSource(ctx context.Context, host string, port int32, heartbeatInterval float64, stopReplicationBefore bool, startReplicationAfter bool) error {
+	m.setReplicationSourceCalls++
+	return nil
+}
+
+func (m *catchupReplicationMysqlDaemon) StopReplication(ctx context.Context, hookExtraEnv map[string]string) error {
+	m.stopReplicationCalls++
+	return nil
+}
+
+func testCatchupPosition(pos uint64) replication.Position {
+	return replication.Position{GTIDSet: replication.FilePosGTID{File: "source-bin.000001", Pos: pos}}
+}

